Index: webrtc/modules/audio_device/ios/objc/RTCAudioSession+Private.h
diff --git a/webrtc/modules/audio_device/ios/objc/RTCAudioSession+Private.h b/webrtc/modules/audio_device/ios/objc/RTCAudioSession+Private.h
index 523a4c10941fcf9a414c473bb798fb6915cf7bc2..fb311853b79afab72c83db3d0dd524b405ba3fe2 100644
--- a/webrtc/modules/audio_device/ios/objc/RTCAudioSession+Private.h
+++ b/webrtc/modules/audio_device/ios/objc/RTCAudioSession+Private.h
@@ -14,9 +14,6 @@ NS_ASSUME_NONNULL_BEGIN
 
 @interface RTCAudioSession ()
 
-/** The lock that guards access to AVAudioSession methods. */
-@property(nonatomic, strong) NSRecursiveLock *lock;
-
 /** The delegates. */
 @property(nonatomic, readonly) NSSet *delegates;
 
