Index: webrtc/modules/modules.gyp
diff --git a/webrtc/modules/modules.gyp b/webrtc/modules/modules.gyp
index 5deef3596ccb986ec7e9847ddb7d5d273abe4dbc..89af024bc60aa24fff86b1dc94c10d314c650517 100644
--- a/webrtc/modules/modules.gyp
+++ b/webrtc/modules/modules.gyp
@@ -471,9 +471,13 @@
               ],
             }],
             ['OS=="ios"', {
+              'includes': [
+                '../build/objc_common.gypi',
+              ],
               'sources': [
                 'video_coding/codecs/h264/h264_video_toolbox_nalu_unittest.cc',
                 'audio_device/ios/audio_device_unittest_ios.cc',
+                'audio_device/ios/objc/RTCAudioSessionTest.mm',
               ],
               # This needs to be kept in sync with modules_unittests.isolate.
               'mac_bundle_resources': [
