Index: webrtc/test/run_all_unittests.cc
diff --git a/webrtc/test/run_all_unittests.cc b/webrtc/test/run_all_unittests.cc
index 486df5ea8cc956ba2a984a36df90f7c77862b163..80166c3ca8a1bd2db395f5e3f095a01cd5e68969 100644
--- a/webrtc/test/run_all_unittests.cc
+++ b/webrtc/test/run_all_unittests.cc
@@ -8,9 +8,13 @@
  *  be found in the AUTHORS file in the root of the source tree.
  */
 
+#include "webrtc/base/scoped_autorelease_pool.h"
 #include "webrtc/test/test_suite.h"
 
 int main(int argc, char** argv) {
+#if defined(WEBRTC_MAC)
+  rtc::ScopedAutoreleasePool pool;
+#endif
   webrtc::test::TestSuite test_suite(argc, argv);
   return test_suite.Run();
 }
