Index: webrtc/webrtc_examples.gyp
diff --git a/webrtc/webrtc_examples.gyp b/webrtc/webrtc_examples.gyp
index 74162a1c8dd876b9c6cf9887f82385df3c774c9a..3baa6901818cac0cd140e7f382b601ca13d265b3 100755
--- a/webrtc/webrtc_examples.gyp
+++ b/webrtc/webrtc_examples.gyp
@@ -281,14 +281,16 @@
           'conditions': [
             ['OS=="ios"', {
               'mac_bundle_resources': [
+                'examples/objc/AppRTCDemo/ios/resources/Roboto-Regular.ttf',
                 'examples/objc/AppRTCDemo/ios/resources/iPhone5@2x.png',
                 'examples/objc/AppRTCDemo/ios/resources/iPhone6@2x.png',
                 'examples/objc/AppRTCDemo/ios/resources/iPhone6p@3x.png',
-                'examples/objc/AppRTCDemo/ios/resources/Roboto-Regular.ttf',
                 'examples/objc/AppRTCDemo/ios/resources/ic_call_end_black_24dp.png',
                 'examples/objc/AppRTCDemo/ios/resources/ic_call_end_black_24dp@2x.png',
                 'examples/objc/AppRTCDemo/ios/resources/ic_clear_black_24dp.png',
                 'examples/objc/AppRTCDemo/ios/resources/ic_clear_black_24dp@2x.png',
+                'examples/objc/AppRTCDemo/ios/resources/ic_surround_sound_black_24dp.png',
+                'examples/objc/AppRTCDemo/ios/resources/ic_surround_sound_black_24dp@2x.png',
                 'examples/objc/AppRTCDemo/ios/resources/ic_switch_video_black_24dp.png',
                 'examples/objc/AppRTCDemo/ios/resources/ic_switch_video_black_24dp@2x.png',
                 'examples/objc/AppRTCDemo/ios/resources/mozart.mp3',
