Index: webrtc/base/base.gyp
diff --git a/webrtc/base/base.gyp b/webrtc/base/base.gyp
index f43fde215c78c9e74629c3fd16306a6b20d472cd..5219cecf5ced97f1abd382da861f63a53602be65 100644
--- a/webrtc/base/base.gyp
+++ b/webrtc/base/base.gyp
@@ -118,6 +118,8 @@
         'ratetracker.cc',
         'ratetracker.h',
         'refcount.h',
+        'scoped_autorelease_pool.h',
+        'scoped_autorelease_pool.mm',
         'scoped_ptr.h',
         'scoped_ref_ptr.h',
         'stringencode.cc',
@@ -322,8 +324,6 @@
         'rollingaccumulator.h',
         'rtccertificate.cc',
         'rtccertificate.h',
-        'scoped_autorelease_pool.h',
-        'scoped_autorelease_pool.mm',
         'scopedptrcollection.h',
         'sec_buffer.h',
         'sha1.cc',
