Index: webrtc/base/objc/RTCDispatcher.h
diff --git a/webrtc/base/objc/RTCDispatcher.h b/webrtc/base/objc/RTCDispatcher.h
index c32b93d47206815bc49fb9ea3b27370a0a33555e..a9b92a661fd1f6c82b90ffff4e23e691993f693c 100644
--- a/webrtc/base/objc/RTCDispatcher.h
+++ b/webrtc/base/objc/RTCDispatcher.h
@@ -16,6 +16,8 @@ typedef NS_ENUM(NSInteger, RTCDispatcherQueueType) {
   // Used for starting/stopping AVCaptureSession, and assigning
   // capture session to AVCaptureVideoPreviewLayer.
   RTCDispatcherTypeCaptureSession,
+  // Used for operations on AVAudioSession.
+  RTCDispatcherTypeAudioSession,
 };
 
 /** Dispatcher that asynchronously dispatches blocks to a specific
