Index: webrtc/base/objc/RTCDispatcher.m
diff --git a/webrtc/base/objc/RTCDispatcher.m b/webrtc/base/objc/RTCDispatcher.m
index 065705a4aea70e368658f4d2e346ef902a6d5d99..3d469c6dab782d48475aa8f5ce4a8ee7235e8bbd 100644
--- a/webrtc/base/objc/RTCDispatcher.m
+++ b/webrtc/base/objc/RTCDispatcher.m
@@ -10,15 +10,20 @@
 
 #import "RTCDispatcher.h"
 
+static dispatch_queue_t kAudioSessionQueue = nil;
 static dispatch_queue_t kCaptureSessionQueue = nil;
 
 @implementation RTCDispatcher {
   dispatch_queue_t _captureSessionQueue;
+  dispatch_queue_t _audioSessionQueue;
 }
 
 + (void)initialize {
   static dispatch_once_t onceToken;
   dispatch_once(&onceToken, ^{
+    kAudioSessionQueue = dispatch_queue_create(
+        "org.webrtc.RTCDispatcherAudioSession",
+        DISPATCH_QUEUE_SERIAL);
     kCaptureSessionQueue = dispatch_queue_create(
         "org.webrtc.RTCDispatcherCaptureSession",
         DISPATCH_QUEUE_SERIAL);
@@ -39,6 +44,8 @@ static dispatch_queue_t kCaptureSessionQueue = nil;
       return dispatch_get_main_queue();
     case RTCDispatcherTypeCaptureSession:
       return kCaptureSessionQueue;
+    case RTCDispatcherTypeAudioSession:
+      return kAudioSessionQueue;
   }
 }
 
