Index: webrtc/examples/objc/AppRTCDemo/ios/ARDVideoCallView.h
diff --git a/webrtc/examples/objc/AppRTCDemo/ios/ARDVideoCallView.h b/webrtc/examples/objc/AppRTCDemo/ios/ARDVideoCallView.h
index 378281d0057311444b1ec0a9eef37e24fdd7c0ae..636443afdf0b3915a1cb88cf23c817ac339bfe9a 100644
--- a/webrtc/examples/objc/AppRTCDemo/ios/ARDVideoCallView.h
+++ b/webrtc/examples/objc/AppRTCDemo/ios/ARDVideoCallView.h
@@ -21,6 +21,9 @@
 // Called when the camera switch button is pressed.
 - (void)videoCallViewDidSwitchCamera:(ARDVideoCallView *)view;
 
+// Called when the route change button is pressed.
+- (void)videoCallViewDidChangeRoute:(ARDVideoCallView *)view;
+
 // Called when the hangup button is pressed.
 - (void)videoCallViewDidHangup:(ARDVideoCallView *)view;
 
